### ISSUE TITLE
SimpleNodeParser should produce TextNodes with deterministic IDs

### ISSUE THREAD

**Issue by g12i** [run-llama/LlamaIndexTS#639]
When initializing the `VectorStoreIndex` with the default service context, it utilizes the `SimpleNodeParser` that divides documents into a sequence of nodes. However, with each call, new nodes are generated with random UUIDs associated. In my view, it would be preferable for this process to be idempotent. Assigning node IDs derived from the SOURCE ID and text hash should be sufficient.

This way, if I call `index.insert` with the same origin Document it won't clutter my storage with ghost nodes.

```diff
// packages/core/src/nodeParsers/utils.ts

  textSplits.forEach((textSplit) => {
    const node = new TextNode({
+     id_: `${document.id_}/${quickHash(textSplit)}`,
      text: textSplit,
      metadata: includeMetadata ? _.cloneDeep(document.metadata) : {},
      excludedEmbedMetadataKeys: _.cloneDeep(
        document.excludedEmbedMetadataKeys,
      ),
      excludedLlmMetadataKeys: _.cloneDeep(document.excludedLlmMetadataKeys),
    });
    node.relationships[NodeRelationship.SOURCE] = document.asRelatedNodeInfo();
    nodes.push(node);
  });
  
+ function quickHash(str: string): string {
+   let hash = 0;
+   for (let i = 0, len = str.length; i < len; i++) {
+     const chr = str.charCodeAt(i);
+     hash = (hash << 5) - hash + chr;
+     hash |= 0; // Convert to 32bit integer
+   }
+   return String(hash);
+ }

```

**Comment by marcusschiesser** [comment]
@g12i Thanks a lot for pointing this out. I think we can not generally add this, as there are different use cases to cover. For example, your `quickHash` value would be the same if the metadata is changing.

But after calling the `SimpleNodeParser`, the split `TextNodes` should have the same `hash` value (if metadata and content are the same).

Are you currently calling `VectorStoreIndex.insertNodes` to add new nodes? This method does not check the `hash` value. 

Have you tried using our [IngestionPipelines](https://ts.llamaindex.ai/modules/ingestion_pipeline/#connecting-to-vector-databases) - they are handling `hash` values.

Alternatively, we could also add an `idFunc` to the `SimpleNodeParser` that generates an ID, and you could pass in your code a customized ID function.

Please let me know what you think, and I will be happy to solve this problem.

**Comment by g12i** [comment]
Hi @marcusschiesser , thanks for your response. 

I'm currently using `VectorStoreIndex.fromDocuments`, which creates nodes using `runTransformations` with `SimpleNodeParser` and thus generating unique IDs for each node. 

```  
VectorStoreIndex.fromDocuments()
└── args.nodes = await runTransformations(documents)
    └── VectorStoreIndex.init()
        └── VectorStoreIndex.buildIndexFromNodes()
            └── VectorStoreIndex.insertNodes()
                └── VectorStoreIndex.insertNodesToStore()
                    ├── vectorStore.add(nodes)
                    └── docStore.addDocuments([nodeWithoutEmbedding])
  ```

I guess I should never call `fromDocuments` more than once with the same Document in the first place, right?

**Comment by marcusschiesser** [comment]
@g12i you might want to try this PR: https://github.com/run-llama/LlamaIndexTS/pull/646 - it checks for duplicates in the `VectorStoreIndex.fromDocuments` function

**Comment by marcusschiesser** [comment]
@g12i will reopen this ticket as we haven't yet merged #646.

**Comment by g12i** [comment]
Thanks @marcusschiesser, I thought it's no longer relevant as you've found different solution to my issue.
